kubectl: fix a panic when createGeneratedObject failed

mountkin opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR fixed two bugs in the `kubectl run` command.
The first commit fixed a panic that can be reproduced with the following two commands:
```bash
kubectl run --image=alpine hello -- sleep 1d
kubectl run --image=alpine --stdin --tty --attach hello -- /bin/sh
```
When the panic happens, the stack trace is as the following: 
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x12f92e6]

goroutine 1 [running]:
k8s.io/kubernetes/pkg/kubectl/cmd.RunRun(0x1890d20, 0xc4207f6210, 0x184e700, 0xc42000c010, 0x184e720, 0xc42000c018, 0x184e720, 0xc42000c020, 0xc42052ef00, 0xc42054f980, ...)
	/home/star/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/run.go:328 +0x10b6
k8s.io/kubernetes/pkg/kubectl/cmd.NewCmdRun.func1(0xc42052ef00, 0xc42054f980, 0x2, 0x8)
	/home/star/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/run.go:105 +0x144
...
```
The reason for the panic is that the original code didn't handle the error returned by `createGeneratedObject` correctly. See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/run.go#L281
If the function `createGeneratedObject` fails, the returned `runObject` will be `nil`.  At https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/run.go#L328 a nil pointer dereference will happen.

The second problem this PR fixed is a resource leak if both `--rm` and `--expose` flags are provided. The original code used a map (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/run.go#L280) to trace the created objects. But at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/run.go#L296 the code will overwrite the previously created object with the same key. So at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/run.go#L348 only the service will be deleted. The deployment will be left behind.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Fixed a panic in `kubectl run --attach ...` when the api server failed to create the runtime object (due to name conflict, PSP restriction, etc.)
```
/kind bug
/sig cli

CaoShuFeng commented on pkg/kubectl/cmd/run.go:
```diff
@@ -276,23 +275,19 @@ func RunRun(f cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *c
-		if err != nil {
-			allErrs = append(allErrs, err)
-		} else {
+		if err == nil {
```
We should deal with the error if `err != nil`, rather than ignore it.

mountkin replied:
Good catch. I didn't aware that the `err` was scoped.
I also fixed another problem in the `run` command.

mountkin committed:
kubectl: fix a panic when createGeneratedObject failed

Reproduce:
kubectl run --image=alpine hello -- sleep 1d
kubectl run --image=alpine --stdin --tty --attach hello -- /bin/sh

Signed-off-by: Shijiang Wei <mountkin@gmail.com>

mountkin committed:
avoid resource leak when both `--rm` and `--expose` are specified

Before this patch, the deployment `leak-test` won't be removed
after the following command exits.
kubectl run --image=nginx:alpine --stdin --attach \
  --rm --expose --port 80 leak-test -- sh
This patch ensures both the deployment and the service are all removed.

Signed-off-by: Shijiang Wei <mountkin@gmail.com>

mountkin commented:
/test pull-kubernetes-integration
/assign @janetkuo

mengqiy commented:
The change LGTM.
Please explain the root cause it the PR description or link to a issue describing it.

mountkin commented:
@mengqiy I have updated the PR description. PTAL again. Thanks

mengqiy commented on pkg/kubectl/cmd/run.go:
```diff
@@ -276,14 +276,14 @@ func RunRun(f cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *c
 	runObject, err := createGeneratedObject(f, cmd, generator, names, params, cmdutil.GetFlagString(cmd, "overrides"), namespace)
 	if err != nil {
-		allErrs = append(allErrs, err)
+		return err
```
You should do
```
allErrs = append(allErrs, err)
utilerrors.NewAggregate(allErrs)
```
You don't want to lose earlier error info.

mountkin replied:
I moved the allErrs variable after the ‘if’ statement. Because this error can be returned immediately, there’s no need to use the error aggregator.

mengqiy commented:
Thanks for your update.

We should exit earlier to fix the 1st issue.

The second issue is not a real problem, since golang has GC. I'm not worrying about it. :)

janetkuo commented:
> The second issue is not a real problem, since golang has GC. I'm not worrying about it. :)

I think it's a real issue. In the map, the service and object share the same map key, and when removing objects only the last one inserted into the map will get removed. 

To reproduce it:
```
$ kubectl run busybox -it --rm --port 5000 --expose=true --image=busybox --restart=Never
```
You'll see that only the service is removed, but the pod is not.

janetkuo commented on pkg/kubectl/cmd/run.go:
```diff
@@ -276,14 +276,14 @@ func RunRun(f cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *c
-		allErrs = append(allErrs, err)
+		return err
 	} else {
-		runObjectMap[generatorName] = runObject
```
The key should be `serviceGenerator`?  
cc @xiangpengzhao

mountkin replied:
I found the key is not used at all,  so I changed it to a slice

xiangpengzhao replied:
@janetkuo yeah it should be `serviceGenerator` and using `serviceGenerator` fixes the second issue. However, as @mountkin mentioned, the key is not used at all. So it's safe to change it to a slice and it does fix the second issue as well.

xiangpengzhao commented:
>To reproduce it:
$ kubectl run busybox -it --rm --port 5000 --expose=true --image=busybox --restart=Never
You'll see that only the service is removed, but not the pod.

I can't believe this issue does exist! I remember I tested this case against my original patch then (I can't recall the testing details though). I verify on my env that both  https://github.com/kubernetes/kubernetes/pull/61713#discussion_r178193880 by @janetkuo and @mountkin 's patch can fix this issue. :+1: Thank you!

janetkuo commented:
/lgtm

simonswine commented:
Just hit that on kubectl 1.10 with a single command. 

```
kubectl --as=developer run busybox-priv --image=busybox --restart=Never -n developer --overrides '{"spec":{"containers":[{"name":"busybox","image":"busybox","command":["uname","-a"],"securityContext":{"privileged":true}}]}}' --attach --rm
``` 

Can we someone add a cherry-pick candidate and a set the milestone. Happy to generate a PR then

simonswine commented:
I can confirm the backport fixes my issues as well (with properly handling the API server error)

MaciekPytel commented:
@mountkin Can you add a release note to this PR? @simonswine added one to cherry-pick, but release notes are generated from base PRs and it would be useful to let users know that this bug was fixed in 1.10.1.

mountkin commented:
@MaciekPytel I updated the pr description and added the release note. Please take a look when you have time. Thanks.

MaciekPytel commented:
Great, thank you!
